pkg/api/types.go

```diff
@@ -1546,6 +1564,120 @@ func (p *Properties) GetCustomCloudIdentitySystem() string {
 	return AzureADIdentitySystem
 }
 
+// SetCustomCloudProfileEnvironment retrives the endpoints from Azure Stack metadata endpoint and populate the Environment
+func (cs *ContainerService) SetCustomCloudProfileEnvironment() {
+	p := cs.Properties
+	if p.IsAzureStackCloud() {
+		env := p.CustomCloudProfile.Environment
+
+		if env.ServiceManagementEndpoint == "" || env.ActiveDirectoryEndpoint == "" || env.GraphEndpoint == "" || env.ResourceManagerVMDNSSuffix == "" {
+			metadataURL := fmt.Sprintf("%s/metadata/endpoints?api-version=1.0", strings.TrimSuffix(env.ResourceManagerEndpoint, "/"))
+
+			// Retrieve the metadata
+			httpClient := &http.Client{
+				Timeout: 15 * time.Second,
+			}
+			endpointsresp, err := httpClient.Get(metadataURL)
+			if err != nil || endpointsresp.StatusCode != 200 {
+				log.Fatalf("%s . apimodel invalid: failed to retrieve Azure Stack endpoints from %s", err, metadataURL)
```

Comment:
The error message may be misleading. A non-200 HTTP response is not always caused by a bad URL.

Reply:
The same comment applies for the next 3 validations. The root cause may not be an `invalid apimodel`.

Reply:
it aligins with other existing error message.

Reply:
I like consistency as well. I'll rather be extra careful and avoid confusing users or CSS as they can end up barking at the wrong tree.

Reply:
agree, I would like to have PM reviewed all the message we have in the furture and make the change together.